### ISSUE TITLE
Problem with MPI_Status conversion when C has 4-byte ints and Fortran has 8-byte INTEGERs

### ISSUE THREAD

**Issue by ahaichen** [open-mpi/ompi#7918]
### Open MPI version: 4.0.1 
- installed from source
- build with: ./configure F77=gfortran FC=gfortran  FFLAGS=-fdefault-integer-8 FCFLAGS=-fdefault-integer-8 
- Operating system: Linux version 4.9.0-12-amd64 (debian-kernel@lists.debian.org)
- CPU: Intel(R) Xeon(R)

Hi all, 

I get an issue when running in 64-bit MPI with the rank initialized 0, the code is 

```fortran
Program test
USE MPI_F08
IMPLICIT NONE

TYPE(MPI_Status)                            :: MPI_stat
Integer(kind=MPI_INTEGER_KIND) :: MPI_err
Integer(kind=MPI_INTEGER_KIND) :: MPI_id=0
Integer(kind=MPI_INTEGER_KIND) :: MPI_np
Integer                                              :: tempint,i_MPI

CALL MPI_Init(MPI_err)
CALL MPI_Comm_rank(MPI_COMM_WORLD,MPI_id,MPI_err)
CALL MPI_Comm_size(MPI_COMM_WORLD,MPI_np,MPI_err)

if(MPI_id==0) tempint=1
if(MPI_id/=0)  tempint=-1

write(*,*) 'MPI_np check1: MPI_id=',MPI_id,'MPI_np=',MPI_np,'tempint',tempint
if(MPI_id/=0) call MPI_Send(tempint,1,MPI_integer8,0,MPI_id,MPI_COMM_WORLD,MPI_err)
DO i_MPI=1,MPI_np-1
  if(MPI_id==0) call MPI_Recv(tempint,1,MPI_integer8,i_MPI,i_MPI,MPI_COMM_WORLD,MPI_stat,MPI_err)
ENDDO
write(*,*) 'MPI_np check2: MPI_id=',MPI_id,'MPI_np=',MPI_np,'tempint',tempint

CALL MPI_Finalize(MPI_err)
End program test
```

It is compiled with 

mpifort -fdefault-integer-8 test.f90

run with 

mpirun -np 3 ./a.out

the output for 3 threads will be 

```
 MPI_np check1: MPI_id=                    0 MPI_np=                    3 tempint                    1
 MPI_np check1: MPI_id=                    2 MPI_np=                    3 tempint                   -1
 MPI_np check1: MPI_id=                    1 MPI_np=                    3 tempint                   -1
 MPI_np check2: MPI_id=                    1 MPI_np=                    3 tempint                   -1
 MPI_np check2: MPI_id=                    0 MPI_np=                    0 tempint                   -1
 MPI_np check2: MPI_id=                    2 MPI_np=                    3 tempint                   -1
```

It seems by giving "MPI_id=0", the "MPI_np" will turn to 0 (should be 3)  on the rank 0 after a MPI_recv call.  
```
MPI_np check2: MPI_id=                    0 MPI_np=                    0 tempint                   -1
```
I do not understand how does this happen. Does anyone have an idea on this issue?  thanks a lot ~

**Comment by jsquyres** [comment]
Two questions:

1. What threads are you referring to?  I do not see any threaded code in your example.
2. I see you compiled your app with `-fdefault-integer-8`.  Did you build Open MPI with the same flag?

Additionally, can you supply the information requested by the github issue template?

* Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
* If you are building/installing from a git clone, please copy-n-paste the output from git submodule status.
* Please describe the system on which you are running
    * Operating system/version:
    * Computer hardware:
    * Network type:

Thanks!

**Comment by ahaichen** [comment]
> Two questions:
> 
> 1. What threads are you referring to?  I do not see any threaded code in your example.
> 2. I see you compiled your app with `-fdefault-integer-8`.  Did you build Open MPI with the same flag?
> 
> Additionally, can you supply the information requested by the github issue template?
> 
> * Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
> * If you are building/installing from a git clone, please copy-n-paste the output from git submodule status.
> * Please describe the system on which you are running
>   
>   * Operating system/version:
>   * Computer hardware:
>   * Network type:
> 
> Thanks!

Hi jsquyres,

I have update my question with these informations. thanks a lot ~

**Comment by rhc54** [comment]
I still don't see any sign that your code is threaded - are you using "thread" to refer to a distinct "process"? Would you please show the cmd line you use to generate that output?

**Comment by ahaichen** [comment]
> I still don't see any sign that your code is threaded - are you using "thread" to refer to a distinct "process"? Would you please show the cmd line you use to generate that output?

sorry, I mean the process, I have updated the question. Thanks ~

**Comment by rhc54** [comment]
I guess I don't understand your question, then. You are _receiving_ the sent values into the ```tempint``` location inside rank=0, and all the rank > 0 procs are sending you a value of -1. So after the loop, of course ```tempint``` in rank=0 is -1.

What am I missing?

**Comment by ahaichen** [comment]
> I guess I don't understand your question, then. You are _receiving_ the sent values into the `tempint` location inside rank=0, and all the rank > 0 procs are sending you a value of -1. So after the loop, of course `tempint` in rank=0 is -1.
> 
> What am I missing?

The result of tempint is correct, but if you check the value "MPI_np" on 0 process, it is 0, which should be 3. 
```
MPI_np check2: MPI_id=                    0 MPI_np=                    0 tempint                   -1
```
Thanks ~

**Comment by jsquyres** [comment]
`F77` and `FFLAGS` aren't used any more; you need to use `FCFLAGS` (and `FC`, which you're already using).

What might be happening here is that you're effectively building Open MPI without `-fdefault-integer-8`, but then building your app with `-fdefault-integer-8`.  Things then go downhill from there because of integer size mismatches between your app and the Open MPI fortran library.

**Comment by ahaichen** [comment]
> `F77` and `FFLAGS` aren't used any more; you need to use `FCFLAGS` (and `FC`, which you're already using).
> 
> What might be happening here is that you're effectively building Open MPI without `-fdefault-integer-8`, but then building your app with `-fdefault-integer-8`. Things then go downhill from there because of integer size mismatches between your app and the Open MPI fortran library.

sorry, I just lost the `FCFLAGS=-fdefault-integer-8` part when i copy the text.  

when I remove the `=0` in  `MPI_id=0`  in the code, the result will be fine. That really confused me.

**Comment by jsquyres** [comment]
Your test program doesn't compile:

```
$ mpifort foo.f90 -o foo
foo.f90:19:42:

   19 |   if(MPI_id/=0) call MPI_Send(tempint,Int1,MPI_integer8,0,MPI_id,MPI_COMM_WORLD,MPI_err)
      |                                          1
Error: Symbol ‘int1’ at (1) has no IMPLICIT type
```

I'm assuming that's supposed to be 1.  Even after fixing that, it does not compile for me:

```
$ mpifort foo.f90 -o foo
foo.f90:19:85:

   19 | ll MPI_Send(tempint,1,MPI_integer8,0,MPI_id,MPI_COMM_WORLD,MPI_err)
      |                                                                   1

Error: There is no specific subroutine for the generic ‘mpi_send’ at (1)
foo.f90:21:100:

   21 | tempint,1,MPI_integer8,i_MPI,i_MPI,MPI_COMM_WORLD,MPI_stat,MPI_err)
      |                                                                   1

Error: There is no specific subroutine for the generic ‘mpi_recv’ at (1)
```

So I'm not sure how you're getting results...?

Can you produce a consistent question that we can validate and test?

Thanks!

**Comment by ahaichen** [comment]
> Your test program doesn't compile:
> 
> ```
> $ mpifort foo.f90 -o foo
> foo.f90:19:42:
> 
>    19 |   if(MPI_id/=0) call MPI_Send(tempint,Int1,MPI_integer8,0,MPI_id,MPI_COMM_WORLD,MPI_err)
>       |                                          1
> Error: Symbol ‘int1’ at (1) has no IMPLICIT type
> ```
> 
> I'm assuming that's supposed to be 1. Even after fixing that, it does not compile for me:
> 
> ```
> $ mpifort foo.f90 -o foo
> foo.f90:19:85:
> 
>    19 | ll MPI_Send(tempint,1,MPI_integer8,0,MPI_id,MPI_COMM_WORLD,MPI_err)
>       |                                                                   1
> 
> Error: There is no specific subroutine for the generic ‘mpi_send’ at (1)
> foo.f90:21:100:
> 
>    21 | tempint,1,MPI_integer8,i_MPI,i_MPI,MPI_COMM_WORLD,MPI_stat,MPI_err)
>       |                                                                   1
> 
> Error: There is no specific subroutine for the generic ‘mpi_recv’ at (1)
> ```
> 
> So I'm not sure how you're getting results...?
> 
> Can you produce a consistent question that we can validate and test?
> 
> Thanks!

Yes, that should be 1, I do compile the code. I get similar problem before on another machine. If you try 

```fortran
Program test
USE MPI_F08
IMPLICIT NONE

TYPE(MPI_Status)                            :: MPI_stat
Integer(kind=MPI_INTEGER_KIND) :: MPI_err
Integer(kind=MPI_INTEGER_KIND) :: MPI_id=0
Integer(kind=MPI_INTEGER_KIND) :: MPI_np
Integer                                              :: tempint,i_MPI

CALL MPI_Init(MPI_err)
CALL MPI_Comm_rank(MPI_COMM_WORLD,MPI_id,MPI_err)
CALL MPI_Comm_size(MPI_COMM_WORLD,MPI_np,MPI_err)

if(MPI_id==0) tempint=1
if(MPI_id/=0)  tempint=-1

write(*,*) 'MPI_np check1: MPI_id=',MPI_id,'MPI_np=',MPI_np,'tempint',tempint
if(MPI_id/=0) call MPI_Send(tempint,INT(1,kind=MPI_INTEGER_KIND),MPI_integer8,&
INT(0,kind=MPI_INTEGER_KIND),MPI_id,MPI_COMM_WORLD,MPI_err)
DO i_MPI=1,MPI_np-1
   if(MPI_id==0) call MPI_Recv(tempint,INT(1,kind=MPI_INTEGER_KIND),MPI_integer8, &
                                i_MPI,i_MPI,MPI_COMM_WORLD,MPI_stat,MPI_err)
ENDDO
write(*,*) 'MPI_np check2: MPI_id=',MPI_id,'MPI_np=',MPI_np,'tempint',tempint

CALL MPI_Finalize(MPI_err)
End program test
```
I guess it should be fine. 

mpifort -fdefault-integer-8 test.f90
mpirun -np 3 ./a.out
```
 MPI_np check1: MPI_id=                    0 MPI_np=                    3 tempint                    1
 MPI_np check1: MPI_id=                    1 MPI_np=                    3 tempint                   -1
 MPI_np check2: MPI_id=                    1 MPI_np=                    3 tempint                   -1
 MPI_np check1: MPI_id=                    2 MPI_np=                    3 tempint                   -1
 MPI_np check2: MPI_id=                    2 MPI_np=                    3 tempint                   -1
 MPI_np check2: MPI_id=                    0 MPI_np=                    0 tempint                   -1
```
Thanks ~

**Comment by ggouaillardet** [comment]
Thanks for the report, this is indeed a genuine bug.

The variable is overwritten at

```
1,1,test (foo.f90:23)
1,1,  mpi_recv_f08 (recv_f08.F90:27)
1,1,    ompi_recv_f (precv_f.c:91)
1,1,      PMPI_Status_c2f (pstatus_c2f.c:98)
```

One option is we should be using (after implementing!) `PMPI_Status_c2f08()` instead of `PMPI_Status_c2f()`
An other option is the defintion of `type(MPI_Status)` is incorrect when `INTEGER` size is `8`

@jsquyres any ideas?

**Comment by ahaichen** [comment]
> Thanks for the report, this is indeed a genuine bug.
> 
> The variable is overwritten at
> 
> ```
> 1,1,test (foo.f90:23)
> 1,1,  mpi_recv_f08 (recv_f08.F90:27)
> 1,1,    ompi_recv_f (precv_f.c:91)
> 1,1,      PMPI_Status_c2f (pstatus_c2f.c:98)
> ```
> 
> One option is we should be using (after implementing!) `PMPI_Status_c2f08()` instead of `PMPI_Status_c2f()`
> An other option is the defintion of `type(MPI_Status)` is incorrect when `INTEGER` size is `8`
> 
> @jsquyres any ideas?

sure, thanks. 

If it is indeed a bug,  Do I have to go back to `USE MPI` at present to make it through? thanks ~

**Comment by jsquyres** [comment]
Ah, I see my mistake -- I did not compile the app with `-fdefault-integer-8` -- All the extra KIND notations in the app confused/distracted me, and I neglected to use that flag.  FWIW, my simpler form of the test program was:

```fortran
Program test
  USE MPI_F08
  IMPLICIT NONE

  TYPE(MPI_Status) :: MPI_stat
  Integer :: rank
  Integer :: size
  Integer :: tempint,i

  CALL MPI_Init()
  CALL MPI_Comm_rank(MPI_COMM_WORLD,rank)
  CALL MPI_Comm_size(MPI_COMM_WORLD,size)

  tempint = -1
  if (rank == 0) tempint=1

  write(*,*) 'MPI_np check1: MPI_id=',rank,'MPI_np=',size,'tempint',tempint
  if (rank == 0) then
     do i = 1, size-1
        call MPI_Recv(tempint, 1, MPI_integer8, i, i, MPI_COMM_WORLD, MPI_stat)
    enddo
  else
     call MPI_Send(tempint, 1, MPI_integer8, 0, rank, MPI_COMM_WORLD)
  endif
  write(*,*) 'MPI_np check2: MPI_id=',rank,'MPI_np=',size,'tempint',tempint

  CALL MPI_Finalize()
End program test
```

@ahaichen No, the problem is not `use mpi` vs. `use mpi_f08`.  The problem is in Open MPI's handling of the difference in default size of C integers vs. Fortran integers (i.e., 4 vs. 8).  @ggouaillardet is pointing out that have an error in converting between C and Fortran status, because of this difference.  I should note that this is not a common case (where size of C int != size of fortran INTEGER), and therefore something that probably has not been tested in years.

An alternative is to build Open MPI and your app without `-fdefault-integer-8` and just increase the size of integers to 8 bytes in your Fortran code as necessary (i.e., just for specific values that need to be 8 byte integers).

@ggouaillardet and I are going to need to confer about this; it may actually be a large issue.

**Comment by jsquyres** [comment]
Note: this probably probably goes back a very long time.  This is a fairly uncommon case, though -- it's likely only worthwhile to fix in v4.0.x and forward.

**Comment by devreal** [comment]
There is an old issue on a similar topic (referencing here as potentially related): https://github.com/open-mpi/ompi/issues/27

**Comment by ahaichen** [comment]
@jsquyres sure, I get it. thanks a lot ~

So, I guess, in principle, it means at present it is impossible to send an extremely long array (>32bit integer) in Fortran.  of course it should be avoided generally.

**Comment by jsquyres** [comment]
> So, I guess, in principle, it means at present it is impossible to send an extremely long array (>32bit integer) in Fortran. of course it should be avoided generally.

I'm not quite sure what you are asking.

Are you asking about sending/receiving an array which is more than 2^31 elements in length?  If so, you can construct a datatype that can be used to send/receive a single message that is longer than 2^31 elements.  E.g., make a TYPE_CONTIGUOUS datatype of 1024 MPI_INTEGERs.  Then you can send/receive up to 2^31 of these, which lets you send/receive arrays up to 2^41 total elements.  Extrapolate that into whatever you need.

Note that MPI-4 will make this situation a little easier (i.e., allowing "count" parameters for send/receive/etc. that are larger than 32 bits in size).  The publication of the MPI-4 standard is still many months away; I don't know when we'll have that implementation available in Open MPI.

(but yes, sending/receiving such huge arrays should be avoided when possible 😄 )

Or are you asking about sending/receiving integer elements that are larger than 32 bits?  If so, you can just use MPI_INTEGER8 for that.

**Comment by ahaichen** [comment]
@jsquyres   yes, I mean > 2^31 elements. thanks for your advice ~

**Comment by jsquyres** [comment]
You could also just do multiple sends/receives.  E.g., if your array is (4 * 2^31) elements in length, just do 4 sends / 4 receives.

The loss in efficiency of 4 sends vs. 1 send in this case is negligible (because the messages are incredibly long).

**Comment by ahaichen** [comment]
Yes, indeed. thanks ~

**Comment by jsquyres** [comment]
@ggouaillardet I see a number of things that could be problems.

## Faulty configury computation

This configury code for determining the value of `OMPI_FORTRAN_STATUS_SIZE` (i.e., what becomes the Fortran constant `MPI_STATUS_SIZE`) seems to be wrong.  In particular, the computation for `num_integers` -- shouldn't that be dividing by the size of a Fortran integer (not `$ac_cv_sizeof_int`)?

https://github.com/open-mpi/ompi/blob/27d30f3d17acc476b4716e2d34c992f9b30941f2/config/ompi_setup_mpi_fortran.m4#L245-L262

If I'm right, this should be relatively straightforward to fix.

## Missing C functions

It looks like we are missing several C functions -- how on earth did that happen?  It's a little embarrassing that I drew figure 17.1 in MPI-3.1 and then didn't even include these functions in Open MPI.  😳 

* `MPI_Status_c2f08()`
* `MPI_Status_f082c()`
* `MPI_Status_f2f08()`
* `MPI_Status_f082f()`

Do you have the cycles to give a shot at implementing these functions, perchance?

(although we have to make some decisions about them -- see below)

## Differing types for MPI_Status

It looks like we have three different types for MPI_Status:

1. The C type, which is
https://github.com/open-mpi/ompi/blob/27d30f3d17acc476b4716e2d34c992f9b30941f2/ompi/include/mpi.h.in#L417-L432
2. The `mpif.h` / `use mpi` type, which is just an INTEGER array of size `MPI_STATUS_SIZE`.
3. The `mpi_f08` type, which is
https://github.com/open-mpi/ompi/blob/27d30f3d17acc476b4716e2d34c992f9b30941f2/ompi/mpi/fortran/use-mpi-f08/mod/mpi-f08-types.F90#L73-L79

I think that once the `OMPI_FORTRAN_STATUS_SIZE` computation is fixed, the C and `mpif.h` statuses should be ok.

But we have a different sized struct over on the `mpi_f08` side -- it has 3 INTEGERs, a C_INT, and a C_SIZE_T.  This is... problematic.

We call a common back-end Fortran function for most (all?) Fortran bindings -- e.g., all the Fortran interfaces for `MPI_Recv()` end up calling `ompi_recv_f()`.  This innermost function (e.g., `ompi_recv_f()`) doesn't know from which outer-level Fortran binding it was invoked: it doesn't know if the status is an `mpif.h` status or an `mpi_f08` status.

The question then becomes: how to make the status assignment correct (or "safe enough") in this innermost function, regardless of whether the Status is from `mpif.h` or `mpi_f08`?

We could fundamentally re-architect our Fortran bindings to always pass this binding information into the innermost Fortran functions (so that we can do the final C-to-[FILL IN THE RIGHT FORTRAN STATUS TYPE]).  This seems like an incredibly invasive solution, and probably very error prone.

The other option would be just to make the `mpi_f08` status be sure to be "big enough" such that the existing Status assignment we do at the bottom of these innermost Fortran routines will be safe.

The `mpif.h` status is `MPI_STATUS_SIZE` INTEGERs long.  The `mpi_f08` `TYPE(MPI_Status)` has fixed integers at the front.  What if we put an array of `(MPI_STATUS_SIZE-3)` INTEGERs at the end?  The canceled/count values are never accessed from Fortran -- see the big comment here: https://github.com/open-mpi/ompi/blob/27d30f3d17acc476b4716e2d34c992f9b30941f2/ompi/mpi/c/status_c2f.c#L75-L96

I.e., something like this:

```fortran
   type, BIND(C) :: MPI_Status
      integer :: MPI_SOURCE
      integer :: MPI_TAG
      integer :: MPI_ERROR
      integer OMPI_PRIVATE :: internal(MPI_STATUS_SIZE-3)
   end type MPI_Status
```

This would make the `mpif.h` and `mpi_f08` statuses be the same size, and we should be good.  Right...?

**Comment by jsquyres** [comment]
@ggouaillardet My initial analysis was close but not exactly right.  See #7921 for a PR that fixes the 4-vs-8 issue (but doesn't add the missing functions).

**Comment by ggouaillardet** [comment]
@ahaichen The current MPI standard specifies the number of elements to be sent/received is an `int` in `C`, and an `INTEGER` in Fortran.
Internally, the `Fortran` bindings use the `C` bindings, so if your intent is to send/receive a number of elements that does not fit in 31 bits, building Open MPI with 64 bits Fortran integers will **not** do the trick.
(internally, the number of elements will be truncated to 32 bits and results into an undefined behavior).

As pointed by @jsquyres, the canonical way to work around this limitation is to use large derived datatypes in order to decrease the number of elements to be sent (you cannot send 2^40 elements in one shot, but you can send 2^10 vectors of 2^30 elements in one shot).
You can also have a look at https://github.com/jeffhammond/BigMPI (note this is not an implementation of the MPI standard strictly speaking), or simply wait for MPI-4 that will likely have a solution for this.

**Comment by ahaichen** [comment]
@ggouaillardet  sure, thanks. is there any way to build it correctly with 64 bit fortran to avoid this problem?  thanks

**Comment by ggouaillardet** [comment]
the current status is that `mpi_f08` bindings are broken when 64 bits Fortran integers are used.
the obvious workaround is to `use mpi`.
you can also avoid using `type(MPI_Status)` and use `MPI_STATUS_IGNORE` instead.

**Comment by jsquyres** [comment]
The benefits of `mpi_f08` outweigh using `mpi`, IMNSHO.

My advice: use `mpi_f08`, and if you need to send giant messages, just use multiple sends or use a datatype.

**Comment by jsquyres** [comment]
This issue was resolved long ago, but we neglected to close this issue:

1. We fixed the issue in what will become Open MPI v5.0.x.  
2. The fix unfortunately involved breaking ABI (because it increased the size of `TYPE(MPI_Status)`.  This is ok to do at Open MPI v5.0, but is not ok to do in v4.0.x and v4.1.x.
2. Therefore, the fix will be available in Open MPI v5.0.x, but in v4.0.x and v4.1.x, we now disallow building Open MPI with sizeof(int)!=sizeof(INTEGER).
